### ISSUE TITLE
Allow arbitrary extrapolation methods

### ISSUE THREAD

**Issue by bergfried** [python-pillow/Pillow#703]
Whenever a function needs to extrapolate values of non-existing pixels (i.e. pixels outside of the image), it should be possible to specify the extrapolation method to use.

The extrapolation methods I consider to be most useful are:
- replicate border
- replicate border, with alpha set to 0
- arbitrary constant
- wrap around
- arbitrary Python function

See the introduction to http://docs.opencv.org/modules/imgproc/doc/filtering.html for further explanation and other extrapolation methods. There are several Pillow functions where this comes in handy. For example,

```python
image.transform(size, PIL.Image.AFFINE, data, PIL.Image.BICUBIC, extrapolate=REPLICATE_BORDER_ALPHA0)
```

might produce a rotated and scaled version of the source image on top of a transparent background so you can paste this transformed image into other images easily. Another example:

```python
image.filter(filter, extrapolate=WRAP_AROUND)
```

might produce a blurred image that you can use for a tiled wallpaper without ugly tile borders.

If you need to prioritize, I consider `Image.transform` to be more important than `Image.filter`. (This is because it is quite easy to work around this issue for `Image.filter` as soon as this is implemented for `Image.transform`.)

**Comment by radarhere** [comment]
> I might be able to submit a pull request to at least support transparent fill, if not extrapolate.

Transparent fill should already be the behaviour.

```python
from PIL import Image, ImageTransform
im = Image.open("Tests/images/hopper.png").convert("RGBA")
transform = ImageTransform.ExtentTransform((0, 0, 150, 150))
im.transform((100, 100), transform).show()
```
gives this image (it's hard to see because it is transparent, but there is transparency to the right and bottom of the image).

![transform](https://user-images.githubusercontent.com/3112309/201468340-d5a684f7-d94e-40c1-bd3e-6ce3d58a0738.png)

**Comment by Yay295** [comment]
Of course, it's only transparent if the image has an alpha channel. Otherwise it's black.

Also there are two different parameters to some of the methods: `fill` and `fillcolor`. Some of the methods take one, or the other, or both, so maybe that could be something to make more consistent.
